knowledge-base/adr/0032-monorepo-tooling.md

```diff
@@ -0,0 +1,131 @@
+- [Powerful filters](https://turborepo.org/docs/features/filtering) allow us to run commands in all relevant packages.
+
+### Changesets
+
+Caveats:
+
+- Need to maintain changeset files (instead of `CHANGELOG.md`).
+- We need special permission to use the bot and need to verify if we can use the action. We could work around this, but this is extra effort.
+- Our current versioning strategy means we have fixed versions for all packages.
+  Changesets would support linked versions, which would avoid meaningless version increases for our packages.
+  We should reconsider our versioning policy given this new opportunity.
+- It maintains one `CHANGELOG.md` per package, but allows custom changelog formatting.
+- When creating a changeset, affected packages need to be selected.
+  This adds some extra effort.
+  - There is a "changed packages" selection, that makes it easy to select all packages that were touched since the last changeset.
+
+<img width="560" alt="changeset console output" src="https://user-images.githubusercontent.com/26565533/164425056-581e4cd6-6863-4d8c-b78f-c13ceabf3b23.png">
+
+Benefits:
+
+- Documentation is good but contains many TODOs.
+- CLI is amazingly simple:
+  - Create a changeset with `yarn changeset`.
+  - Apply changeset with `yarn changeset version`.
+  - Create release with `yarn changeset publish` or `yarn changeset publish --tag beta`.
+- Each release needs to have one or more changesets.
+- Changesets are merged so you could add a changeset every commit if needed.
+- Changesets give space to describe [what, why, and how](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md#i-am-in-a-multi-package-repository-a-mono-repo) for every change.
+- Adopted successfully in several prominent monorepo projects.
+- There is active development backed by a committed third party (Atlassian).
+- Independent versioning is possible.
+
+Release pipeline SHOULD work, but is not tested as of 21st of April and some adjustments should be expected.
+
+## Consequences
+
+### Changelog
+
+- Changelog needs to be maintained via the `yarn changeset` CLI instead of manually writing in the `CHANGELOG.md`.
+- Each package will have its own `CHANGELOG.md` which will be generated by Changesets.
+- Review of changelog files may be a bit more cumbersome due to it being split into multiple files. Alternatively it can be done during each PR.
+
+### Adjusting scripts
+
+- Scripts and their dependencies need to be maintained in the `turbo.json` file.
+- Errors in this config may lead to unexpected behavior.
+
+### Calling scripts in a workspace
+
+- As dependencies between scripts may only be considered when using `turbo run SCRIPT --filter=my-package`, we should avoid using `yarn workspace mypackage SCRIPT`.
+
+### Finding errors
+
+- As scripts are executed in parallel, it may be slightly harder to find error logs (though it still should be fairly easy).
```

Comment:
Is there a run in band option for debugging?

Reply:
Yes, you can use `--concurrency=1` (or `--serial` but this is deprecated). Is there a place I should document this?

Reply:
perhaps here as a hint for finding errors.